python/lsst/daf/butler/registry/queries/_results.py

```diff
@@ -709,3 +762,152 @@ def explain_no_results(self) -> Iterator[str]:
+
+        Notes
+        -----
+        This counts the number of rows returned, not the number of unique rows
+        returned, so even with ``exact=True`` it may provide only an upper
+        bound on the number of *deduplicated* result rows.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def any(self, *, execute: bool = True, exact: bool = True) -> bool:
+        """Test whether this query returns any results.
+
+        Parameters
+        ----------
+        execute : `bool`, optional
+            If `True`, execute at least a ``LIMIT 1`` query if it cannot be
+            determined prior to execution that the query would return no rows.
+        exact : `bool`, optional
+            If `True`, run the full query and perform post-query filtering if
+            needed, until at least one result row is found.  If `False`, the
+            returned result does not account for post-query filtering, and
+            hence may be `True` even when all result rows would be filtered
+            out.
+
+        Returns
+        -------
+        any : `bool`
+            `True` if the query would (or might, depending on arguments) yield
+            result rows.  `False` if it definitely would not.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def order_by(self, *args: str) -> DimensionRecordQueryResults:
+        """Make the iterator return ordered result.
+
+        Parameters
+        ----------
+        *args : `str`
+            Names of the columns/dimensions to use for ordering. Column name
+            can be prefixed with minus (``-``) to use descending ordering.
+
+        Returns
+        -------
+        result : `DataCoordinateQueryResults`
+            Returns ``self`` instance which is updated to return ordered
+            result.
+
+        Notes
+        -----
+        This method can modify the iterator in place and return the same
+        instance.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def limit(self, limit: int, offset: Optional[int] = None) -> DimensionRecordQueryResults:
+        """Make the iterator return limited number of records.
+
+        Parameters
+        ----------
+        limit : `int`
+            Upper limit on the number of returned records.
+        offset : `int` or `None`
+            If not `None` then the number of records to skip before returning
+            ``limit`` records.
+
+        Returns
+        -------
+        result : `DataCoordinateQueryResults`
+            Returns ``self`` instance which is updated to return limited set
+            of records.
+
+        Notes
+        -----
+        This method can modify the iterator in place and return the same
+        instance. Normally this method is used together with `order_by`
+        method.
+        """
+        raise NotImplementedError()
+
+
+class DatabaseDimensionRecordQueryResults(DimensionRecordQueryResults):
+    """
+    """
```

Comment:
Missing docstring.

You're also welcome to merge this in with the ABC (or not), as far as I'm concerned.  I do think we want an ABC _somewhere_ to distinguish between query results objects from `RemoteRegistry` vs. `SqlRegistry`, but I sort of feel like the right level for that is in a helper object held by a concrete+final `QueryResults` object, not the `QueryResults` object itself.  But I'm also not at all sure either way.